### PR TITLE
Storage implementation

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: src/storage/filestorage.hpp [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+#define FILE_STORAGE_HPP_
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <tinycrypt/constants.h>
+#include <tinycrypt/sha256.h>
+#include <unistd.h>
+
+#include "aos/common/tools/error.hpp"
+
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/filestorage.hpp]: Done

File: src/storage/filestorage.hpp [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+    {
+        auto ret = lseek(mFd, sizeof(Header), SEEK_SET);
+        if (ret < 0) {
+            return AOS_ERROR_WRAP(errno);
+        }
+
+        Record  record {};
+        ssize_t nread {};
+
+        while ((nread = read(mFd, &record, sizeof(Record))) == sizeof(Record)) {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/filestorage.hpp:91]: Done

File: src/storage/filestorage.hpp [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+            if (record.mDeleted) {
+                continue;
+            }
+
+            if (filter(record.mData)) {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/filestorage.hpp]: Done

File: src/storage/storage.cpp [hunk 4]
```diff
@@ -16,25 +18,196 @@ aos::Error Storage::Init()
...
+    LOG_DBG() << "Storage directory: " << cStoragePath;
+
+    auto instancePath = aos::FS::JoinPath(cStoragePath, "instance.db");
+    LOG_DBG() << "Instance database path: " << instancePath;
+    if (!(err = mInstanceDatabase.Init(instancePath)).IsNone()) {
+        return err;
+    }
+
+    auto servicePath = aos::FS::JoinPath(cStoragePath, "service.db");
+    LOG_DBG() << "Service database path: " << servicePath;
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/storage.cpp:38]: Done

File: src/storage/storage.cpp [hunk 5]
```diff
@@ -16,25 +18,196 @@ aos::Error Storage::Init()
...
+    auto err = mInstanceDatabase.ReadRecords([&instances, this](const InstanceInfo& instanceInfo) -> aos::Error {
+        auto instance = ConvertInstanceInfo(instanceInfo);
+        auto err = instances.PushBack(instance);
+        if (!err.IsNone() && !err.Is(aos::ErrorEnum::eNoMemory)) {
+            return err;
+        }
+
+        return aos::ErrorEnum::eNone;
+    });
+
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/storage.cpp:92]: Done

File: src/storage/storage.cpp [hunk 6]
```diff
@@ -16,25 +18,196 @@ aos::Error Storage::Init()
...
+    if (!err.IsNone() && !err.Is(aos::ErrorEnum::eNoMemory)) {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/storage.cpp]: Done

File: src/storage/storage.cpp [hunk 7]
```diff
@@ -16,25 +18,196 @@ aos::Error Storage::Init()
...
+    if (!err.IsNone()) {
+        return aos::RetWithError<aos::sm::servicemanager::ServiceData>(aos::sm::servicemanager::ServiceData {}, err);
+    }
+
+    return aos::RetWithError<aos::sm::servicemanager::ServiceData>(ConvertServiceData(serviceData));
+}
+
+Storage::InstanceInfo Storage::ConvertInstanceInfo(const aos::InstanceInfo& instance)
+{
+    InstanceInfo instanceInfo;
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/storage.cpp:155]: Done

File: src/storage/storage.cpp [hunk 8]
```diff
@@ -16,25 +18,196 @@ aos::Error Storage::Init()
...
+    strcpy(instanceInfo.mStatePath, instance.mStatePath.CStr());
+    strcpy(instanceInfo.mStoragePath, instance.mStoragePath.CStr());
+    instanceInfo.mUID = instance.mUID;
+
+    return instanceInfo;
+}
+
+aos::InstanceInfo Storage::ConvertInstanceInfo(const InstanceInfo& instance)
+{
+    aos::InstanceInfo instanceInfo;
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/storage.cpp:170]: Done

File: src/storage/storage.cpp [hunk 9]
```diff
@@ -16,25 +18,196 @@ aos::Error Storage::Init()
...
+    instanceInfo.mStatePath = instance.mStatePath;
+    instanceInfo.mStoragePath = instance.mStoragePath;
+    instanceInfo.mUID = instance.mUID;
+
+    return instanceInfo;
+}
+
+Storage::ServiceData Storage::ConvertServiceData(const aos::sm::servicemanager::ServiceData& service)
+{
+    ServiceData serviceData;
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/storage.cpp:185]: Done

File: src/storage/storage.cpp [hunk 10]
```diff
@@ -16,25 +18,196 @@ aos::Error Storage::Init()
...
+    strcpy(serviceData.mServiceID, service.mServiceID.CStr());
+    strcpy(serviceData.mProviderID, service.mProviderID.CStr());
+    strcpy(serviceData.mImagePath, service.mImagePath.CStr());
+
+    return serviceData;
+}
+
+aos::sm::servicemanager::ServiceData Storage::ConvertServiceData(const Storage::ServiceData& service)
+{
+    aos::sm::servicemanager::ServiceData serviceData;
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/storage.cpp:199]: Done

File: src/storage/storage.hpp [hunk 11]
```diff
@@ -52,6 +58,86 @@ class Storage : public aos::sm::launcher::StorageItf, private aos::NonCopyable {
...
+    /**
+     * Returns service data by service ID.
+     *
+     * @param serviceID service ID.
+     * @return  RetWithError<ServiceData>.
+     */
+    aos::RetWithError<aos::sm::servicemanager::ServiceData> GetService(const aos::String& serviceID) override;
+
+private:
+    constexpr static const char* cStoragePath = CONFIG_AOS_STORAGE_DIR;
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/storage.hpp]: Done

**Review by Eabramov84 — APPROVED** [review]

File: src/storage/storage.cpp [hunk 12]
```diff
@@ -16,25 +18,196 @@ aos::Error Storage::Init()
...
+    if (!(err = aos::FS::MakeDirAll(cStoragePath)).IsNone()) {
+        LOG_ERR() << "Failed to create storage directory: " << err;
+
+        return err;
+    }
+
+    LOG_DBG() << "Storage directory: " << cStoragePath;
+
+    auto instancePath = aos::FS::JoinPath(cStoragePath, "instance.db");
+    LOG_DBG() << "Instance database path: " << instancePath;
```
[the remark this answers is not in the document]
Comment (al1img) [reply src/storage/storage.cpp:32]: I don't see it is changed.

File: src/storage/storage.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (al1img) [reply src/storage/storage.cpp:38]: The same,.

[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/storage.cpp]: Sorry, done

File: src/storage/storage.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/storage.cpp]: Sorry, done

**Review by al1img — APPROVED** [review]

**Pull request merged**